[Tooling] Use bot for release management git operations

iangmaia opened the pull request:
## References

Android counterpart: https://github.com/wordpress-mobile/WordPress-Android/pull/21028
Original task describing the problem: https://github.com/Automattic/apps-infra-plans/issues/39
See paaHJt-6EP-p2 for more details.

## What it does
This PR updates the release pipelines to use the `use-bot-for-git` script to use a bot account for git operations, avoiding the need for SSH deploy keys with write permission.

**Note**: this PR also upgrades Xcode to `16.1-rc.1`.

### How to test

We'll fully run the release pipelines in the next release cycle once this is merged.

To test it, I've commented out a few side effects from `start_code_freeze` on 79a7f89daccc0188d4123f428fa8c77b933a223b and ran it directly on Buildkite (setting the env var `PIPELINE=release-pipelines/code-freeze.yml`) with [this resulting build](https://buildkite.com/automattic/wordpress-ios/builds/24280).

iangmaia committed:
Update pipelines to use `use-bot-for-git` script for running git operations with write access

dangermattic commented:
<!--
  0 Errors
  1 Warning: PR is not assigned to a milest...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is not assigned to a milestone.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iangmaia committed:
[TO REVERT] Comment out side effects to `code_freeze` lane, keeping only version bump

wpmobilebot committed:
Bump version number

iangmaia committed:
Revert "[TO REVERT] Comment out side effects to `code_freeze` lane, keeping only version bump"

This reverts commit 3580a5241af37ea20987381f5b5ae2956a401066.

iangmaia committed:
Revert "Bump version number"

This reverts commit 5ae6bfd616ec825145f3d6d2abf4dde7a2193824.

iangmaia committed:
Remove specific simulators

mokagio commented on .buildkite/commands/configure-for-release.sh:
```diff
@@ -1,13 +1,16 @@
+if [[ "${BASH_SOURCE[0]}" == "${0}" ]]; then
+  echo "This script must be 'source'd (instead of being called directly as an executable) to work properly"
+  exit 1
+fi
```
When a script is meant to be sourced, the shebang is ignored. (I learned that the hard way, but it's mentioned [in Wikipedia](https://en.wikipedia.org/wiki/Shebang_(Unix)#cite_note-9:~:text=The%20shebang%20line%20is%20usually%20ignored%20by%20the%20interpreter%2C%20because%20the%20%22%23%22%20character%20is%20a%20comment%20marker%20in%20many%20scripting%20languages%3B%20some%20language%20interpreters%20that%20do%20not%20use%20the%20hash%20mark%20to%20begin%20comments%20still%20may%20ignore%20the%20shebang%20line%20in%20recognition%20of%20its%20purpose.%5B9%5D) and there are StackOverflow questions like [this one](https://stackoverflow.com/questions/49080559/source-script-does-not-honor-bin-bash-ex-and-exit-when-a-command-fails) about it.)

That's all to say, we should replicate the flags manually after ensuring the script is sourced.

```suggestion
if [[ "${BASH_SOURCE[0]}" == "${0}" ]]; then
  echo "This script must be 'source'd (instead of being called directly as an executable) to work properly"
  exit 1
fi

set -e
set -u
```

Or... we might remove them entirely and rely on the caller to manage that? 🤔 ...Nah, I think being extra cautious here is best.

mokagio commented on .buildkite/commands/configure-for-release.sh:
```diff
@@ -1,13 +1,16 @@
-# Buildkite is currently using the HTTPS URL to checkout.
-# We need to override it to be able to use the deploy key.
-git remote set-url origin git@github.com:wordpress-mobile/WordPress-iOS.git
+echo '--- :robot_face: Use bot for git operations'
```
Nitpick
```suggestion
echo '--- :robot_face: Use bot for Git operations'
```

mokagio commented on .buildkite/commands/configure-for-release.sh:
```diff
@@ -1,13 +1,16 @@
-curl -L https://api.github.com/meta | jq -r '.ssh_keys | .[]' | sed -e 's/^/github.com /' >> ~/.ssh/known_hosts
+add_host_to_ssh_known_hosts github.com
 git config --global user.email "mobile+wpmobilebot@automattic.com"
 git config --global user.name "Automattic Release Bot"
```
Following up on @AliSoftware's suggestion from one of our private repos, I think if we moved these three line in `use-bot-for-git` would then allow us to call the command directly without having to wrap it in this script, right?

iangmaia replied:
I've updated the agents and the code here accordingly: 477184657d7cc0e541bb5f9e781a282b4af724fd (Buildkite test run [here](https://buildkite.com/automattic/wordpress-ios/builds/24295)).

mokagio reviewed (approved):
Looking good, only nitpicks.

iangmaia committed:
Apply suggestions from code review

Co-authored-by: Gio Lodi <gio@mokacoding.com>

iangmaia committed:
Add comment to configure-for-release.sh explaining the separate flags

iangmaia committed:
Use `use-bot-for-git` directly
The `use-bot-for-git` script in the mac agents will now setup the bot Git user, so doing that isn't necessary anymore
